src/Tasks/WriteCodeFragment.cs

```diff
@@ -245,5 +274,312 @@ private string GenerateCode(out string extension)
+                        }
+                        else
+                        {
+                            // For named parameters, use the type of the property if we can find it.
+                            value = ConvertParameterValueToInferredType(
+                                attributeType.Value?.GetProperty(parameter.Name)?.PropertyType,
+                                parameter.Value,
+                                parameter.Name
+                            );
+                        }
+
+                        break;
+
+                }
+
+                attribute.Arguments.Add(new CodeAttributeArgument(parameter.Name, value));
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Finds the types that the parameters are likely to be, by finding a constructor
+        /// on the attribute that has the same number of parameters that have been provided.
+        /// Returns an array of types with a length equal to the number of positional parameters.
+        /// If no suitable constructor is found, the array will contain null types.
+        /// </summary>
+        private Type[] FindPositionalParameterTypes(Type attributeType, IReadOnlyList<AttributeParameter> positionalParameters)
+        {
+            // The attribute type might not be known.
+            if (attributeType is not null)
+            {
+                List<Type[]> candidates;
+
+                // Find the constructors with the same number
+                // of parameters as we will be specifying.
+                candidates = attributeType
+                    .GetConstructors()
+                    .Select(c => c.GetParameters().Select(p => p.ParameterType).ToArray())
+                    .Where(t => t.Length == positionalParameters.Count)
+                    .ToList();
+
+                if (candidates.Count == 1)
+                {
+                    return candidates[0];
+                }
+                else if (candidates.Count > 1)
+                {
+                    Log.LogMessageFromResources("WriteCodeFragment.MultipleConstructorsFound");
+
+                    // Before parameter types could be specified, all parameter values were
+                    // treated as strings. To be backward-compatible, we need prefer the
+                    // constructor that has all string parameters, if it exists.
+                    var allStringParameters = candidates.FirstOrDefault(c => c.All(t => t == typeof(string)));
+
+                    if (allStringParameters is not null)
+                    {
+                        return allStringParameters;
+                    }
+
+                    // There isn't a constructor where all parameters are strings, so we are free
+                    // to pick one of the candidates. To ensure that we always select the same
+                    // constructor, sort the candidates by type names, then pick the first one.
+                    return candidates
```

Comment:
I don't understand this part. Are there multiple equally valid constructors, as long as there is no constructor of the right length with only string parameters? Is there some way of figuring out which constructor we should prefer?

Reply:
> Is there some way of figuring out which constructor we should prefer?

Possibly, but it felt like overkill given how rarely this code path will be hit. We could:
1. Look at other parameter types that _do_ have declared types and use them to select a constructor, but we're still not guaranteed to find a single constructor if we do that.
2. We could see which type the parameter value "looks like". For example, `42` is numeric, so we could match it to the constructor that has a numeric type for that parameter. But again, we're still not guaranteed to find a single constructor.
3. Just pick one and if it wasn't the correct one, then the consumer will need to declare the type of the parameter. 

I choose option three because it was the simplest, but also because this particular code path is very unlikely to be hit. The reason for that is how the attribute type is found. The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.type.gettype?view=net-5.0#System_Type_GetType_System_String_) for the `Type.GetType(string)` method says this about the `typeName` parameter:

> The assembly-qualified name of the type to get. See `AssemblyQualifiedName`. _If the type is in the currently executing assembly or in mscorlib.dll/System.Private.CoreLib.dll, it is sufficient to supply the type name qualified by its namespace._

Because the attribute's type name needs to be used in code as the name of the attribute, that means an assembly-qualified name _cannot_ be used, which means the only attribute types that we can support are those that are found in this assembly (`Microsoft.Build.Tasks`) or in `mscorlib`. This assembly has no attributes that can be added to assemblies, which means the attributes can only come from `mscorlib`. 

Looking into `mscorlib` we find that:
* There are 216 attributes currently defined in `mscorlib`.
* 92 of those attributes can be added to assemblies.
* Ignoring the default constructor, only 10 of those attributes have more than one constructor.
* Excluding attributes that have a constructor with strings for every parameter leaves us with six attributes.
* And from those attributes, if we only take attributes that have overloads with the same number of parameters, we are left with five attributes.

Those attributes are:
* `System.Reflection.AssemblyAlgorithmIdAttribute`
* `System.Reflection.AssemblyFlagsAttribute`
* `System.Runtime.CompilerServices.CompilationRelaxationsAttribute`
* `System.Runtime.InteropServices.ClassInterfaceAttribute`
* `System.Runtime.InteropServices.IDispatchImplAttribute`

The code I used to find these types is:
```csharp
typeof(string)
    .Assembly
    .GetTypes()
    .Where(t => typeof(Attribute).IsAssignableFrom(t))
    .Where(t => ((AttributeUsageAttribute)t.GetCustomAttribute(typeof(AttributeUsageAttribute))).ValidOn.HasFlag(AttributeTargets.Assembly))
    .Where(t => t.GetConstructors().Where(c => c.GetParameters().Length > 0).Count() > 1)
    .Where(t => !t.GetConstructors().Any(c => c.GetParameters().All(p => p.ParameterType == typeof(string))))
    .SelectMany(t => t.GetConstructors().Select(c => c.GetParameters().Select(p => p.ParameterType.Name)).ToArray(), (t, p) => new { Type = t, Parameters = p})
    .GroupBy(x => new { x.Type, Parameters = x.Parameters.Count() })
    .Where(g => g.Count() > 1)
    .Select(g => new {g.Key.Type, Constructors = g.Select(x => string.Join(", ", x.Parameters))})
```

So for five attribute types, it didn't seem like it was worth the effort to do anything else. I should add a comment that briefly explains this reasoning.

As for why it's ordering them, that ensures that the _same_ constructor is always chosen, regardless of the order that the attributes are returned in from `attributeType.GetConstructors()`.

Reply:
Thank you for the explanation! That sounds reasonable to me. I didn't realize that case affected so few constructors.